src/App/Fossa/BinaryDeps/Archive.hs

```diff
@@ -0,0 +1,34 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+module App.Fossa.BinaryDeps.Archive (withArchive, extractZip) where
+
+import Codec.Archive.Zip qualified as Zip
+import Control.Algebra (Has)
+import Control.Effect.Exception (bracket)
+import Control.Effect.Lift (Lift, sendIO)
+import Path (Abs, Dir, File, Path, fromAbsDir, fromAbsFile, mkRelDir, (</>))
+import Path.IO (getTempDir, removeDirRecur)
+
+-- | Use the provided extraction function to extract the zip, then run the callback.
+-- When the callback is complete, clean up the temporary directory holding the archive contents.
+withArchive :: Has (Lift IO) sig m => (Path Abs File -> Path Abs Dir -> m ()) -> Path Abs File -> (Path Abs Dir -> m c) -> m c
+withArchive extract archive = bracket (extractArchive extract archive) cleanupDir
```

Comment:
This module shares some duplication with `Discovery.Archive`. I opted to extract this functionality because:

* The `Discovery.Archive` functionality seems particularly suited for expanding during walk operations; I wanted something that could more easily be called on on specific files.
* I wanted to be able to return data from within the closure, which `Discovery.Archive` doesn't currently do (and I didn't want to mess with something so central if I could avoid it).

That being said I'm definitely open to refactoring this into something better if anyone has strong opinions and/or suggestions how to overcome these concerns.

Reply:
> and I didn't want to mess with something so central if I could avoid it

In haskell, due to the strong typing, this is often less of a concern.  Although it can be expensive, if you change the types on purpose to break things, you'll know exactly what to change in the rest of the codebase.  Like rust, haskell tries to be a "refactor without fear" language.  Especially the way we use it.

Reply:
Fair enough. I'll try just modifying `Discovery.Archive` instead.